apps/gradient_autoscheduler/GradientAutoscheduler.cpp

```diff
@@ -0,0 +1,1008 @@
+                       AutoSchedulerResults *auto_scheduler_results) {
+    // The first few steps are the same as src/AutoSchedule.cpp
+    // Make an environment map which is used throughout the auto scheduling process.
+    std::map<std::string, Function> env;
+    for (const auto &func : outputs) {
+        std::map<std::string, Function> local_env =
+            find_transitive_calls(func);
+        env.insert(local_env.begin(), local_env.end());
+    }
+    // Compute the topological order
+    std::vector<std::string> top_order = topological_order(outputs, env);
+    // Run a pre-pass that inline all trivial Funcs (i.e. the cost of
+    // computing a Func <= calling that Func).
+    // XXX: Note that the cost is estimated using heuristics based on CPU statistics
+    // so this can be bad on GPU.
+    if (inline_all_trivial_functions(outputs, top_order, env)) {
+        // Recompute env map since some functions are inlined.
+        env.clear();
+        for (Function f : outputs) {
+            std::map<std::string, Function> more_funcs = find_transitive_calls(f);
+            env.insert(more_funcs.begin(), more_funcs.end());
+        }
+    }
+    std::vector<std::string> order =
+        realization_order(outputs, env).first;
+    // Repeatedly inline the functions that are only used by another function
+    while (inline_all_element_wise_functions(outputs, order, env)) {
+        // Recompute env map since some functions are inlined.
+        env.clear();
+        for (Function f : outputs) {
+            std::map<std::string, Function> more_funcs = find_transitive_calls(f);
+            env.insert(more_funcs.begin(), more_funcs.end());
+        }
+        order = realization_order(outputs, env).first;
+    }
+
+    // Bounds inference using the given estimation
+    std::vector<Box> output_bounds_expr;
+    output_bounds_expr.reserve(outputs.size());
+    for (const auto &output : outputs) {
+        const FuncSchedule &schedule = output.schedule();
+        const std::vector<Bound> &estimates = schedule.estimates();
+        user_assert((int)estimates.size() == output.dimensions()) <<
+            "Bound estimates of function " << output.name() << " are not provided.\n";
+        std::vector<Interval> b;
+        b.reserve(estimates.size());
+        for (const auto &e : estimates) {
+            b.push_back(Interval(e.min, simplify(e.min + e.extent - 1)));
+        }
+        output_bounds_expr.push_back(Box(b));
+    }
+
+    std::map<std::string, Box> func_bounds = inference_bounds(outputs, output_bounds_expr);
+    for (const auto &it : func_bounds) {
+        const Box &bounds = it.second;
+        for (int d = 0; d < (int)bounds.size(); d++) {
+            user_assert(bounds[d].is_bounded()) << "Access to function or buffer " << it.first << " at dimension " << d << " is not bounded. "
+                                                << "We can only differentiate bounded accesses.\n";
```

Comment:
This seems like an odd error message -- are we doing a differentiation here that I don't see?

Reply:
Fixed.